### DISCUSSION TITLE
How to enable paste events?

### DISCUSSION THREAD

**Discussion by sonsu-lee** [ueberdosis/tiptap#4118] [General Questions]
I'm using tiptap on a vue3 quasar. and Here's the problem I'm facing

1. Whenever I paste data, it's not getting newlines. It's all showing up as a single line. 

2. Moreover, for table tags, it only turns into SPAN tags. 
=> I suspect this is caused by tiptap not detecting the table tag - I tried adding all the table extensions and still had the same situation.

## Here's what I tried

```javascript
function pasteEventHandler(e) {
        e.preventDefault();
        const text = e.clipboardData.getData('text');
        const lines = text.split(/\r?\n/);
        const html = lines.map(line => `<p>${line}</p>`).join('');
        editor.value.commands.insertContent(html);
        // document.querySelector('.ProseMirror').innerHtml = html;
      }

```

I've registered this function with the paste event of editor-content like this
```javascript
<editor-content
      @paste="pasteEventHandler"
      :editor="editor"
      class="editor-content"
    />
```

This will add the new content in addition to what is already pasted.

So I found a way to disable paste through [this article](https://github.com/ueberdosis/tiptap/issues/456), but the handlePaste function didn't work. 

### Here's how I used it
```javascript
const editor = useEditor({
        handlePaste: {
          ...
        },

        ...
}
```

and I tried [paste-rules](https://tiptap.dev/guide/custom-extensions/#paste-rules), but not work

### I tried like this
```javascript
const editor = useEditor({
     extensions: [
          Paragraph.extend({
            addPasteRules() {
             console.log('test');
              return [
                markPasteRule({
                  find: pasteRegex,
                  type: this.type,
                }),
              ]
          })

        ...
}

```

This extension felt like it only ran once when the editor was rendered. 

I also tried using parseOptions, but that didn't work either. 

```javscript
const editor = useEditor({
        parseOptions: {
          preserveWhitespace: 'full',
        },
        ...
);
```

I don't know what else to try, is the only answer to touch ProseMirror's API directly?
If you need to touch it, I'd be grateful if you could tell me how to approach it in tiptap. 
I'm having a hard time with this.  Thanks

**Comment by nooblyf** [comment]
did you find anything? documentation isn't really helpful on paste rules

**Comment by PhilVargas** [comment]
if you're just trying to disable paste, you can use the handlePaste function. unfortunately the return functionality of this function is not documented anywhere, but if you `return false` from handlePaste it will **allow** the paste to continue, if you `return true` it will **prevent** the paste from continuing.

so you need to make sure at the end of your handlePaste, after you've changed the content to your liking, you `return true` and that will stop the editor from inputting the pasted content.

I have not figured out how to use the `slice` argument in any capacity, and haven't figured out how to reach this function in tests. if i trigger a paste event or cmd+v keydown event, it does not seem to propagate to this function

**Comment by YaSh8202** [comment]
If you want to handle the pasteEvent, 
```
Paragraph.extend({
        addProseMirrorPlugins() {
          const plugin = new Plugin({
            props: {
              handleDOMEvents: {
                paste(view, event) {
                  const items = event.clipboardData.items;

                  pasteHandler(event);
                },
              },
            },
          });

          return [plugin];
        },
```

**Comment by MarcusVirg** [comment]
Building off of @PhilVargas's answer, I found a way to handle paste behavior document-wide.

The basic idea is to use the `handlePaste` function that is exposed by TipTap but note that this is part of the [ProseMirror API documented here](https://prosemirror.net/docs/ref/#view.EditorProps.handlePaste). Return `true` from this function to tell Prose not to apply the slice it created by default. Instead you can build your own slice and then apply the slice yourself in a transaction. Here is an example:

```ts
...otherEditorOptions
editorProps: {
	attributes: {
		class: 'min-h-full'
	},
	handlePaste(view, event, slice) {
		const schema = view.state.schema
		// Convert content to blocks data model
		const rawContent = event.clipboardData?.getData('text/plain')
		// Use `const rawContent = event.clipboardData?.getData('text/html')` if you'd rather deal with HTML
		// Build TipTap JSON content: import type { JSONContent } from '@tiptap/react'
		const content = someDeserializeFunc(rawContent) // i.e. [{ type: 'heading', content: [] }]
		// Convert JSON content to ProseMirror nodes and apply them in a transaction
		const nodes = content.map((c) => Node.fromJSON(schema, c))
		const fragment = Fragment.fromArray(nodes)
		const pastedSlice = new Slice(fragment, slice.openStart, slice.openEnd)
		const pasteTr = view.state.tr.replaceRange(
			view.state.selection.from,
			view.state.selection.to,
			pastedSlice
		)
		view.updateState(view.state.apply(pasteTr))
		return true
	}
},
```

In this example, `someDeserializeFunc` would be a function that runs logic on top of unstructured text that you will have to write yourself depending on the behavior you expect.

I have only tested this on plain text with some simple markdown formatting, not sure how image data could be handled with this method.

Also worth checking out:
- [Slice Documentation](https://prosemirror.net/docs/ref/#model.Slice)
- [Fragment Documentation](https://prosemirror.net/docs/ref/#model.Fragment)
